File: algorithms/154.Find_Minimum_in_Rotated_Sorted_Array_II/Find_Minimum_in_Rotated_Sorted_Array_II.cpp
#include <bits/stdc++.h>

class Solution {
public:
    int findMin(vector<int> &num) {
        int left = 0, right = num.size() - 1;
        int mid, ans = num[0];
        while (left <= right) {
            mid = (left + right) / 2;
            if (num[left] < num[mid]) {
                ans = min(ans, num[left]);
                left = mid + 1;
            } else if (num[right] > num[mid]) {
                ans = min(ans, num[mid]);
                right = mid;
            } else {
                ans = min(ans, num[left]);
                ++left;
            }
        }
        return ans;
    }
};
class Solution {
public:
    int findMin(vector<int> &num) {
        int low = 0;
        int high = num.size()-1;
        while(low < high)
        {
            if(num[low] < num[high])
            //sorted
                return num[low];

            int mid = low + (high-low)/2;   //instead of (low+high)/2 to prevent overflow
            if(num[low] == num[mid])
            {
                if(low+1 == high)
                {//low+1==high
                    if(num[low]>num[high])
                        low = high;
                    break;
                }
                else
                    low ++;
            }
            else if(num[low] < num[mid])
                low = mid;
            else
                high = mid; //mid may be the minimum, thus not mid-1
        }
        return num[low];
    }
};

int main() {
    vector<int> num = {3, 3, 1, 3};
    Solution s;
    cout << s.findMin(num) << endl;
    return 0;
}

File: algorithms/154.Find_Minimum_in_Rotated_Sorted_Array_II/Find_Minimum_in_Rotated_Sorted_Array_II.py
class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def findMin(self, nums):
        return min(nums)
class Solution:
    # @param num: a rotated sorted array
    # @return: the minimum number in the array
    def findMin(self, num):
        # write your code here
        min = num[0]
        start, end = 0, len(num) - 1
        while start<end:
            mid = (start+end)/2
            if num[mid]>num[end]:
                start = mid+1
            elif num[mid]<num[end]:
                end = mid
            else:
                end = end - 1
        return num[start]
class Solution:
    # @param num, a list of integer
    # @return an integer
    def findMin(self, num):
        low, high = 0, len(num)
           
        while low < high - 1 and num[low] >= num[high - 1]:
            mid = low + (high - low) / 2
                
            if num[mid] > num[low]:
                low = mid + 1
            elif num[mid] < num[low]:
                if mid == high - 1:
                    return num[mid]
                else:
                    high = mid + 1
            else:
                low += 1

        return num[low]

class Solution2:
    # @param num, a list of integer
    # @return an integer
    def findMin(self, num):
        low, high = 0, len(num) - 1
           
        while low < high and num[low] >= num[high]:
            mid = low + (high - low) / 2
            
            if num[mid] > num[low]:
                low = mid + 1
            elif num[mid] < num[low]:
                high = mid
            else:
                low += 1

        return num[low]

if __name__ == "__main__":
    print Solution().findMin([3, 1, 1, 2, 2, 3])
    print Solution2().findMin([2, 2, 2, 3, 3, 1])
